src/main/java/org/testng/internal/Parameters.java

```diff
@@ -264,35 +264,79 @@ else if (ITestResult.class.equals(c)) {
                                        MethodParameters params,
                                        XmlSuite xmlSuite) {
     List<Object> vResult = Lists.newArrayList();
-    for (int i = 0, j = 0; i < parameterTypes.length; i++) {
-      if (j < parameterNames.length) {
-        String p = parameterNames[j];
-        String value = params.xmlParameters.get(p);
-        if (null == value) {
-          // try SysEnv entries
-          value = System.getProperty(p);
+    if (optionalValues.length != parameterNames.length) {
+      FilterOutInJectedTypesResult filterOutResult = filterOutInJectedTypesFromOptionalValues(parameterTypes, optionalValues);
+      optionalValues = filterOutResult.getOptionalValues();
+      parameterTypes = filterOutResult.getParameterTypes();
+    }
+    for (int i = 0; i < parameterNames.length; i++) {
+      String p = parameterNames[i];
+      String value = params.xmlParameters.get(p);
+      if (null == value) {
+        // try SysEnv entries
+        value = System.getProperty(p);
+      }
+      if (null == value) {
+        if (optionalValues != null) {
+            value = optionalValues[i];
         }
         if (null == value) {
-          if (optionalValues != null) {
-            value = optionalValues[i];
-          }
-          if (null == value) {
             throw new TestNGException("Parameter '" + p + "' is required by "
                     + methodAnnotation
                     + " on " + prefix + " "
                     + name
                     + " but has not been marked @Optional or defined\n"
                     + (xmlSuite.getFileName() != null ? "in "
-                    + xmlSuite.getFileName() : ""));
-          }
+                            + xmlSuite.getFileName() : ""));
+        }
+    }
+      vResult.add(convertType(parameterTypes[i], value, p));
+    }
+    return vResult;
+  }
+  
+  /**
+   * Remove injected types from parameterTypes and optionalValues
+   * @param parameterTypes
+   * @param optionalValues
+   * @return FilterOutInJectedTypesResult
+   */
+  static FilterOutInJectedTypesResult filterOutInJectedTypesFromOptionalValues(Class<?>[] parameterTypes, String[] optionalValues) {
+    List<Class<?>> typeList = Lists.newArrayList(parameterTypes);
+    List<String> optionalValueList = Lists.newArrayList(optionalValues);
+    Iterator<Class<?>> typeIterator = typeList.iterator();
+    Iterator<String> optionalIterator = optionalValueList.iterator();
+    while (typeIterator.hasNext()) {
+        Class<?> parameterType = typeIterator.next();
+        optionalIterator.next();
+        if (INJECTED_TYPES.contains(parameterType)) {
+            optionalIterator.remove();
```

Comment:
The logic is not easy to understand. Why do you remove things instead of iterating on `parameterTypes` and adding appropriate things to a list?

In a pseudo code, I'd prefer something like:

````java
List<FilterOutInJectedType> result = Lists.newArrayList();
for (int i=0; i < parameterTypes.lenght; i++) {
  Class<?> parameterType = parameterTypes[i];  
  if (!INJECTED_TYPES.contains(parameterType)) {
    String optionalValue = optionalValues[i];
    result.add(new FilterOutInJectedType(parameterType, optionalValue));
  }
}
return result;
```